pkg/asset/installconfig/gcp/resourcetags.go

```diff
@@ -0,0 +1,165 @@
+	"github.com/sirupsen/logrus"
+	"golang.org/x/time/rate"
+	"google.golang.org/api/iterator"
+	"google.golang.org/api/option"
+
+	"github.com/openshift/installer/pkg/types/gcp"
+	"github.com/openshift/installer/pkg/types/gcp/validation"
+)
+
+const (
+	// gcpTagsRequestRateLimit is the tag request rate limit per second.
+	gcpTagsRequestRateLimit = 8
+
+	// gcpTagsRequestTokenBucketSize is the burst/token bucket size used
+	// for limiting API requests.
+	gcpTagsRequestTokenBucketSize = 8
+
+	// tagRequestEndPoint is the endpoint for tag requests.
+	tagRequestEndPoint = "https://cloudresourcemanager.googleapis.com"
+
+	// projectParentPathFmt is the format string for parent path of a project resource
+	projectParentPathFmt = "//cloudresourcemanager.googleapis.com/projects/%s"
+)
+
+// GetUserTags returns user-defined tags as a map with key having the form
+// `tagKeys/{tag_key_id}` and value of the form `tagValues/{tag_value_id}`
+// and returns error when project already has any of the defined tags.
+func (c *Client) GetUserTags(project string, userTags []gcp.UserTag) (map[string]string, error) {
+
+	tags, err := c.getTagValueNames(userTags)
+	if err != nil {
+		return nil, err
+	}
+
+	if err = c.findTagConflicts(project, tags); err != nil {
+		return nil, err
+	}
+
+	return tags, nil
+}
+
+// getTagValueNames returns list of tags as required by terraform APIs fetched
+// using the user provided tag ParentID, Key shortname and Value shortname.
+func (c *Client) getTagValueNames(tags []gcp.UserTag) (map[string]string, error) {
+	if len(tags) == 0 {
+		logrus.Debugf("getTagValuesNames: user-defined tags list is empty")
+		return nil, nil
+	}
+
+	// GCP has a rate limit of 600 requests per minute, restricting
+	// to 8 requests per second.
+	limiter := newLimiter(gcpTagsRequestRateLimit, gcpTagsRequestTokenBucketSize, true)
+
+	client, err := c.getTagValueClient()
+	if err != nil {
+		return nil, errors.Wrapf(err, "getTagValuesNames: failed to create gcp tag value client")
+	}
+	defer client.Close()
+
+	getTagValueReq := rscmgrpb.GetNamespacedTagValueRequest{}
+	tagValues := make(map[string]string, len(tags))
+	for _, tag := range tags {
+		if err := limiter.Wait(context.TODO()); err != nil {
+			return nil, errors.Wrapf(err, "tag values name rate limit wait failed")
+		}
+
+		getTagValueReq.Name = fmt.Sprintf("%s/%s/%s", tag.ParentID, tag.Key, tag.Value)
+		tag, err := client.GetNamespacedTagValue(context.TODO(), &getTagValueReq)
+		if err != nil {
+			return nil, errors.Wrapf(err, "fetching %s tag value name failed", getTagValueReq.Name)
+		}
+		tagValues[tag.Parent] = tag.Name
+	}
+
+	return tagValues, nil
+}
+
+// findTagConflicts returns error when the user-defined tags already existing at the
+// project in which the resources will be created. Tags defined at project will be
+// inherited by the child resources and adding same tag will result in conflict error.
+func (c *Client) findTagConflicts(project string, tags map[string]string) error {
```

Comment:
Conflicts and checking for the existence of tags should be done in the GCP installconfig validation: https://github.com/openshift/installer/blob/master/pkg/asset/installconfig/gcp/validation.go

Ideally this would include unit tests to verify this logic.  

Because this code path is so API intensive, it might be worth adding something like AWS's metadata to hold the tag data: https://github.com/openshift/installer/blob/master/pkg/asset/installconfig/aws/metadata.go

Reply:
> Conflicts and checking for the existence of tags should be done in the GCP installconfig validation: https://github.com/openshift/installer/blob/master/pkg/asset/installconfig/gcp/validation.go
>
> Ideally this would include unit tests to verify this logic.

Suggestion incorporated

> Because this code path is so API intensive, it might be worth adding something like AWS's metadata to hold the tag data: https://github.com/openshift/installer/blob/master/pkg/asset/installconfig/aws/metadata.go

I think I didn't understand why it would be required also I feel it's not required, since there no unknown/dynamic values to pass around, and it will be just for tags. Please let me know WDYT?